<a href="https://colab.research.google.com/github/piusAI/AllAboutDL/blob/main/codej_250719.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
## check YOLO dataset

In [ ]:
!pip install ultralytics
from ultralytics import YOLO

In [ ]:
model=YOLO('yolo11n-pose.pt')

In [ ]:
img='Ji.jpg'

results = model.predict(img, conf=0.8, iou=0.25)
results[0].save_txt('test.txt')


image 1/1 /content/Ji.jpg: 640x480 1 person, 135.2ms
Speed: 9.1ms preprocess, 135.2ms inference, 280.2ms postprocess per image at shape (1, 3, 640, 480)


'test.txt'

In [ ]:
import torch

for result in results:
  obj_num = len(result.boxes)

  if obj_num == 1:

    for i in range(obj_num):
      r_class = result.boxes.cls
      r_xywhn = result.boxes.xywhn
      r_kpts_conf = result.keypoints.conf
      r_kpts_xy = result.keypoints.xyn

      kpts_num = r_kpts_xy.shape[1]
      print(kpts_num)
      r_kpts = []
      for idx in range(kpts_num):
        xy_i = r_kpts_xy[i][idx]
        conf_i = r_kpts_conf[i][idx].unsqueeze(-1)

        kpts_i = torch.cat((xy_i, conf_i), dim=0)
        r_kpts.append(kpts_i)
        print(f'kpt{i+1} : {kpts_i.tolist()}')

      r_kpts_tensor = torch.stack(r_kpts)
      print(f'all kpts for object {obj_idx} : {r_kpts_tensor}')


17
kpt1 : [0.42321082949638367, 0.15455494821071625, 0.975811243057251]
kpt1 : [0.4473299980163574, 0.12922316789627075, 0.8591005802154541]
kpt1 : [0.3883821368217468, 0.1330929398536682, 0.9407287240028381]
kpt1 : [0.4642060101032257, 0.1215243861079216, 0.23742279410362244]
kpt1 : [0.31827032566070557, 0.12892670929431915, 0.7511647343635559]
kpt1 : [0.4863925576210022, 0.19375868141651154, 0.9939571619033813]
kpt1 : [0.22239017486572266, 0.22101514041423798, 0.9797098636627197]
kpt1 : [0.6843876242637634, 0.21151888370513916, 0.9866973757743835]
kpt1 : [0.09319159388542175, 0.3466481566429138, 0.913262665271759]
kpt1 : [0.8683290481567383, 0.21306374669075012, 0.9815940856933594]
kpt1 : [0.1055002510547638, 0.4567333161830902, 0.9367347955703735]
kpt1 : [0.450979083776474, 0.508387565612793, 0.9989234805107117]
kpt1 : [0.29226043820381165, 0.53054279088974, 0.997580885887146]
kpt1 : [0.5513973236083984, 0.7034522891044617, 0.9920925498008728]
kpt1 : [0.4195491075515747, 0.725371122

In [ ]:
# <class> <x> <y> <width> <height> <px1> <py1> <p1_vis> ... <pxn> <pyn> <pn_vis>
import torch
import numpy as np

r_kps=[]
for result in results :
  num_obj = len(result.boxes)
  if num_obj == 1:


    kp_conf = result.keypoints.conf
    kp_xy= result.keypoints.xyn
    num_kps = kp_xy.shape[1]

    for obj_idx in range(num_obj):
      r_class = result.boxes.cls[obj_idx]
      r_x, r_y, r_width, r_height = result.boxes.xywhn[obj_idx]

    for i in range(num_kps):
      xy_i = kp_xy[obj_idx, i]
      conf_i = kp_conf[obj_idx, i]
      print(conf_i)

      conf_i = kp_conf[obj_idx, i].unsqueeze(-1)
      print(conf_i)
      keypoint_i = torch.cat((xy_i, conf_i), dim=0)

      r_kps.append(keypoint_i)
      # print(f'keypoint {i+1} : {keypoint_i.tolist()}')
    r_keypoints_tensor = torch.stack(r_kps)
    # print(f'all keypoints for object {obj_idx} : {r_keypoints_tensor}')

tensor(0.9758, device='cuda:0')
tensor([0.9758], device='cuda:0')
tensor(0.8591, device='cuda:0')
tensor([0.8591], device='cuda:0')
tensor(0.9407, device='cuda:0')
tensor([0.9407], device='cuda:0')
tensor(0.2374, device='cuda:0')
tensor([0.2374], device='cuda:0')
tensor(0.7512, device='cuda:0')
tensor([0.7512], device='cuda:0')
tensor(0.9940, device='cuda:0')
tensor([0.9940], device='cuda:0')
tensor(0.9797, device='cuda:0')
tensor([0.9797], device='cuda:0')
tensor(0.9867, device='cuda:0')
tensor([0.9867], device='cuda:0')
tensor(0.9133, device='cuda:0')
tensor([0.9133], device='cuda:0')
tensor(0.9816, device='cuda:0')
tensor([0.9816], device='cuda:0')
tensor(0.9367, device='cuda:0')
tensor([0.9367], device='cuda:0')
tensor(0.9989, device='cuda:0')
tensor([0.9989], device='cuda:0')
tensor(0.9976, device='cuda:0')
tensor([0.9976], device='cuda:0')
tensor(0.9921, device='cuda:0')
tensor([0.9921], device='cuda:0')
tensor(0.9834, device='cuda:0')
tensor([0.9834], device='cuda:0')
tensor(0.8

=> xy, conf붙힐려했는데 한번 더 감싸야 하는듯!

test.txt:

0 0.507903 0.499475 0.953769 0.993625 0.423211 0.154555 0.975811 0.44733 0.129223 0.859101 0.388382 0.133093 0.940729 0.464206 0.121524 0.237423 0.31827 0.128927 0.751165 0.486393 0.193759 0.993957 0.22239 0.221015 0.97971 0.684388 0.211519 0.986697 0.0931916 0.346648 0.913263 0.868329 0.213064 0.981594 0.1055 0.456733 0.936735 0.450979 0.508388 0.998923 0.29226 0.530543 0.997581 0.551397 0.703452 0.992093 0.419549 0.725371 0.983399 0.623713 0.855018 0.866583 0.508093 0.888994 0.772593

## grok vibe Code

In [ ]:
import torch  # tensor 操作 위해 (Ultralytics가 torch 기반이지만 명시 임포트)
from ultralytics import YOLO  # YOLO 클래스 임포트

# 1. 모델 로드 (pose 모델 – keypoints 추출 위해)
model = YOLO('yolo11n-pose.pt')  # pretrained pose nano. 내부: task='pose' 설정, self.model에 nn.Module 로드.

# 2. 예측 실행 (예시 – 실제로는 cap.read() frame 입력)
# results = model.predict('path/to/image.jpg', conf=0.4, iou=0.45)  # 또는 네 비디오 프레임.
# 가정: results 이미 있음 (네 이전 코드처럼).

# 3. Results 루프 (각 이미지/프레임 처리 – results는 List[Results])
for result in results:  # len(results) 보통 1 (single frame), but batch/multi-image 지원.
    # 객체 수 확인 (num_objects = len(result.boxes))
    num_objects = len(result.boxes)  # e.g., 1 if one person detected.
    if num_objects == 0:
        print("No objects detected.")
        continue  # 객체 없으면 스킵 – keypoints도 없음.

    # keypoints 접근 (result.keypoints: Keypoints 객체)
    kp_xy = result.keypoints.xyn  # shape: (num_objects, num_keypoints, 2) – normalized xy.
    kp_conf = result.keypoints.conf  # shape: (num_objects, num_keypoints) – visibility/conf.
    num_keypoints = kp_xy.shape[1]  # e.g., 17 for COCO.

    # 객체 루프 (multi-object 지원 – 보통 0번 객체)
    for obj_idx in range(num_objects):
        # Class와 BBox 추출 (normalized – result.boxes.xywhn: (num_objects, 4))
        r_class = result.boxes.cls[obj_idx]  # tensor scalar, e.g., 0 (person).
        r_bbox = result.boxes.xywhn[obj_idx]  # tensor (4,): [x_center, y_center, width, height].
        r_x, r_y, r_width, r_height = r_bbox  # unpack – 오타 고침 (r_hegit → r_height).

        print(f'class: {r_class.item()}')  # .item()으로 scalar 변환 (tensor → float/int).
        print(f'x: {r_x.item()}')
        print(f'y: {r_y.item()}')
        print(f'width: {r_width.item()}')
        print(f'height: {r_height.item()}')

        # Keypoints 루프 (각 keypoint i에 대해 xy[i] + conf[i] cat)
        r_keypoints = []  # list로 모든 keypoint 모음 (or tensor stack).
        for i in range(num_keypoints):
            # obj_idx 객체의 i번 keypoint xy: kp_xy[obj_idx, i] – (2,)
            xy_i = kp_xy[obj_idx, i]  # tensor [px, py].

            # conf_i: kp_conf[obj_idx, i] – scalar, unsqueeze(-1)으로 (1,) → dim 맞춤.
            conf_i = kp_conf[obj_idx, i].unsqueeze(-1)  # (1,) tensor.

            # cat: xy_i (2,) + conf_i (1,) → (3,) [px, py, p_vis].
            keypoint_i = torch.cat((xy_i, conf_i), dim=0)  # dim=0: 세로 붙임.

            r_keypoints.append(keypoint_i)  # list 추가 – 나중에 tensor화 if 필요.
            print(f'Keypoint {i+1}: {keypoint_i.tolist()}')  # [px, py, p_vis] 출력 – .tolist()로 list 변환.

        # 전체 keypoints tensor (optional – (num_keypoints, 3))
        r_keypoints_tensor = torch.stack(r_keypoints)  # list → tensor (17, 3).
        print(f'All keypoints for object {obj_idx}: {r_keypoints_tensor}')

class: 0.0
x: 0.39514702558517456
y: 0.5217764377593994
width: 0.454997181892395
height: 0.8452656269073486
Keypoint 1: [0.4967406690120697, 0.20241065323352814, 0.36338236927986145]
Keypoint 2: [0.49129149317741394, 0.18468891084194183, 0.019234465435147285]
Keypoint 3: [0.5005658864974976, 0.1865578591823578, 0.3188028335571289]
Keypoint 4: [0.42790791392326355, 0.17306898534297943, 0.19619373977184296]
Keypoint 5: [0.4887053966522217, 0.18604332208633423, 0.967747688293457]
Keypoint 6: [0.3306882381439209, 0.24151931703090668, 0.9935165643692017]
Keypoint 7: [0.4676341712474823, 0.274423211812973, 0.9988104104995728]
Keypoint 8: [0.22017505764961243, 0.35669535398483276, 0.9323804378509521]
Keypoint 9: [0.5107688903808594, 0.4336088299751282, 0.9908181428909302]
Keypoint 10: [0.17927154898643494, 0.441179096698761, 0.8593749403953552]
Keypoint 11: [0.5803928971290588, 0.5265860557556152, 0.9635399580001831]
Keypoint 12: [0.25884899497032166, 0.49997809529304504, 0.9993973970413208]
